lib/spack/spack/binary_distribution.py

```diff
@@ -334,6 +404,8 @@ def build_tarball(spec, outdir, force=False, rel=False, unsigned=False,
     buildinfo['relative_prefix'] = os.path.relpath(
         spec.prefix, spack.store.layout.root)
     spec_dict['buildinfo'] = buildinfo
+    spec.concretize()
```

Comment:
Is this needed?  I would assume that any spec passed to `build_tarball` is concrete -- probably needs an assert or at least a check and a raised `ValueError` -- or are you relying on the spec to be concretized here?

Reply:
It seemed the spec was not concrete here, and then computing the full_hash in the next line would throw an exception.  As this was the last possible moment before the spec information was written out to the `.spec.yaml`, I assumed that if it hadn't been done by this point, then I should do it myself.